File: js/src/path-test.js
// @flow

import {assert} from 'chai';
import {suite, test} from 'mocha';
import {equals} from './compare.js';

import Path from './path.js';
import type {valueOrPrimitive} from './value.js';
import {newList} from './list.js';
import {newMap} from './map.js';
import {newStruct} from './struct.js';

suite('Path', () => {

  async function assertPathEqual(expect: any, ref: valueOrPrimitive, path: Path):
      Promise<void> {
    // $FlowIssue: need to be able to pass in null for ref
    const actual = await path.resolve(ref);
    if (actual === undefined || expect === undefined) {
      assert.strictEqual(expect, actual);
      return;
    }

    assert.isTrue(equals(expect, actual));
  }

  test('struct', async () => {
    const v = newStruct('', {
      foo: 'foo',
      bar: false,
      baz: 203,
    });

    await assertPathEqual('foo', v, new Path().addField('foo'));
    await assertPathEqual(false, v, new Path().addField('bar'));
    await assertPathEqual(203, v, new Path().addField('baz'));
    await assertPathEqual(undefined, v, new Path().addField('notHere'));

    const v2 = newStruct('', {
      v1: v,
    });

    await assertPathEqual('foo', v2, new Path().addField('v1').addField('foo'));
    await assertPathEqual(false, v2, new Path().addField('v1').addField('bar'));
    await assertPathEqual(203, v2, new Path().addField('v1').addField('baz'));
    await assertPathEqual(undefined, v2, new Path().addField('v1').addField('notHere'));
    await assertPathEqual(undefined, v2, new Path().addField('notHere').addField('foo'));
  });

  test('list', async () => {
    const v = await newList([1, 3, 'foo', false]);

    await assertPathEqual(1, v, new Path().addIndex(0));
    await assertPathEqual(3, v, new Path().addIndex(1));
    await assertPathEqual('foo', v, new Path().addIndex(2));
    await assertPathEqual(false, v, new Path().addIndex(3));
    await assertPathEqual(undefined, v, new Path().addIndex(4));
    await assertPathEqual(undefined, v, new Path().addIndex(-4));
  });

  test('map', async () => {
    const v = await newMap([[1, 'foo'], ['two', 'bar'], [false, 23], [2.3, 4.5]]);

    await assertPathEqual('foo', v, new Path().addIndex(1));
    await assertPathEqual('bar', v, new Path().addIndex('two'));
    await assertPathEqual(23, v, new Path().addIndex(false));
    await assertPathEqual(4.5, v, new Path().addIndex(2.3));
    await assertPathEqual(undefined, v, new Path().addIndex(4));
  });

  test('struct.list.map', async () => {
    const m1 = await newMap([['a', 'foo'], ['b','bar'], ['c', 'car']]);
    const m2 = await newMap([['d', 'dar'], [false, 'earth']]);
    const l = await newList([m1, m2]);
    const s = newStruct('', {
      foo: l,
    });

    await assertPathEqual(l, s, new Path().addField('foo'));
    await assertPathEqual(m1, s, new Path().addField('foo').addIndex(0));
    await assertPathEqual('foo', s, new Path().addField('foo').addIndex(0).addIndex('a'));
    await assertPathEqual('bar', s, new Path().addField('foo').addIndex(0).addIndex('b'));
    await assertPathEqual('car', s, new Path().addField('foo').addIndex(0).addIndex('c'));
    await assertPathEqual(undefined, s, new Path().addField('foo').addIndex(0).addIndex('x'));
    await assertPathEqual(undefined, s, new Path().addField('foo').addIndex(2).addIndex('c'));
    await assertPathEqual(undefined, s, new Path().addField('notHere').addIndex(2).addIndex('c'));
    await assertPathEqual(m2, s, new Path().addField('foo').addIndex(1));
    await assertPathEqual('dar', s, new Path().addField('foo').addIndex(1).addIndex('d'));
    await assertPathEqual('earth', s, new Path().addField('foo').addIndex(1).addIndex(false));
  });

  function assertToString(expect: string, path: Path) {
    assert.strictEqual(expect, path.toString());
  }

  test('toString()', () => {
    assertToString('[0][1][100]', new Path().addIndex(0).addIndex(1).addIndex(100));
    assertToString('["0"]["1"]["100"]',
        new Path().addIndex('0').addIndex('1').addIndex('100'));
    assertToString('.foo[0].bar[4.5][false]',
        new Path().addField('foo').addIndex(0).addField('bar').addIndex(4.5).addIndex(false));
  });
});

File: types/path_test.go
package types

import (
	"testing"

	"github.com/attic-labs/testify/assert"
)

func assertPathResolvesTo(assert *assert.Assertions, expect, ref Value, p Path) {
	if expect == nil {
		assert.Nil(p.Resolve(ref))
		return
	}

	assert.True(expect.Equals(p.Resolve(ref)))
}

func TestPathStruct(t *testing.T) {
	assert := assert.New(t)

	v := NewStruct("", structData{
		"foo": NewString("foo"),
		"bar": Bool(false),
		"baz": Number(203),
	})

	assertPathResolvesTo(assert, NewString("foo"), v, NewPath().AddField("foo"))
	assertPathResolvesTo(assert, Bool(false), v, NewPath().AddField("bar"))
	assertPathResolvesTo(assert, Number(203), v, NewPath().AddField("baz"))
	assertPathResolvesTo(assert, nil, v, NewPath().AddField("notHere"))
}

func TestPathList(t *testing.T) {
	assert := assert.New(t)

	v := NewList(Number(1), Number(3), NewString("foo"), Bool(false))

	assertPathResolvesTo(assert, Number(1), v, NewPath().AddIndex(Number(0)))
	assertPathResolvesTo(assert, Number(3), v, NewPath().AddIndex(Number(1)))
	assertPathResolvesTo(assert, NewString("foo"), v, NewPath().AddIndex(Number(2)))
	assertPathResolvesTo(assert, Bool(false), v, NewPath().AddIndex(Number(3)))
	assertPathResolvesTo(assert, nil, v, NewPath().AddIndex(Number(4)))
	assertPathResolvesTo(assert, nil, v, NewPath().AddIndex(Number(-4)))
}

func TestPathMap(t *testing.T) {
	assert := assert.New(t)

	v := NewMap(
		Number(1), NewString("foo"),
		NewString("two"), NewString("bar"),
		Bool(false), Number(23),
		Number(2.3), Number(4.5),
	)

	assertPathResolvesTo(assert, NewString("foo"), v, NewPath().AddIndex(Number(1)))
	assertPathResolvesTo(assert, NewString("bar"), v, NewPath().AddIndex(NewString("two")))
	assertPathResolvesTo(assert, Number(23), v, NewPath().AddIndex(Bool(false)))
	assertPathResolvesTo(assert, Number(4.5), v, NewPath().AddIndex(Number(2.3)))
	assertPathResolvesTo(assert, nil, v, NewPath().AddIndex(Number(4)))
}

func TestPathMutli(t *testing.T) {
	assert := assert.New(t)

	m1 := NewMap(
		NewString("a"), NewString("foo"),
		NewString("b"), NewString("bar"),
		NewString("c"), NewString("car"),
	)

	m2 := NewMap(
		NewString("d"), NewString("dar"),
		Bool(false), NewString("earth"),
	)

	l := NewList(m1, m2)

	s := NewStruct("", structData{
		"foo": l,
	})

	assertPathResolvesTo(assert, l, s, NewPath().AddField("foo"))
	assertPathResolvesTo(assert, m1, s, NewPath().AddField("foo").AddIndex(Number(0)))
	assertPathResolvesTo(assert, NewString("foo"), s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(NewString("a")))
	assertPathResolvesTo(assert, NewString("bar"), s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(NewString("b")))
	assertPathResolvesTo(assert, NewString("car"), s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(NewString("c")))
	assertPathResolvesTo(assert, nil, s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(NewString("x")))
	assertPathResolvesTo(assert, nil, s, NewPath().AddField("foo").AddIndex(Number(2)).AddIndex(NewString("c")))
	assertPathResolvesTo(assert, nil, s, NewPath().AddField("notHere").AddIndex(Number(0)).AddIndex(NewString("c")))
	assertPathResolvesTo(assert, m2, s, NewPath().AddField("foo").AddIndex(Number(1)))
	assertPathResolvesTo(assert, NewString("dar"), s, NewPath().AddField("foo").AddIndex(Number(1)).AddIndex(NewString("d")))
	assertPathResolvesTo(assert, NewString("earth"), s, NewPath().AddField("foo").AddIndex(Number(1)).AddIndex(Bool(false)))
}

func TestPathToString(t *testing.T) {
	assert := assert.New(t)
	assert.Equal("[0]", NewPath().AddIndex(Number(0)).String())
	assert.Equal("[\"0\"][\"1\"][\"100\"]", NewPath().AddIndex(NewString("0")).AddIndex(NewString("1")).AddIndex(NewString("100")).String())
	assert.Equal(".foo[0].bar[4.5][false]", NewPath().AddField("foo").AddIndex(Number(0)).AddField("bar").AddIndex(Number(4.5)).AddIndex(Bool(false)).String())
}
